### PR TITLE
lsm_local_disk: Add SMART Status query method

### PR DESCRIPTION
**Pull request by joehandzik**
This patch adds the ability to query SMART Status for SAS and SATA drives via the local_disk APIs.

Signed-Off-By: Joe Handzik <joseph.t.handzik@hpe.com>

### CODE REVIEW

**Comment by joehandzik** [comment]
@tasleson @cathay4t Finally got the SMART Status code to a place I was happy with. I have tested this against three variations, effectively:

1. Smart HBA to a SATA drive,
2. Smart HBA to a SAS drive, and
3. ahci controller to a SATA drive

I have intentionally limited the scope of what I'm doing here (only supporting ATA PASSTHROUGH 12 instead of both 12 and 16 and  not supporting 48 bit ATA commands being the big ones IMO), but let me know if you think I've limited too far. I also dumbed down the status that comes back from SAS drives to what ATA can support, but I'm thinking we'll want some sort of advanced SMART Status API to go and get the extra data that SAS can support (it can isolate what is failing, not just tell you that something is failing).

File: c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h [hunk 1]
```diff
@@ -99,6 +99,36 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *disk_path,
                                             char **vpd83,
                                             lsm_error **lsm_err);
 
+/** New in version 1.4.
```
Comment (tasleson) [inline c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h]: 1.4 Has already been released, so this should be 1.5

File: c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h [hunk 2]
```diff
@@ -99,6 +99,36 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *disk_path,
...
+ *                           1 (LSM_DISK_SMART_STATUS_WARN):
+ *                              SMART Status warning.
+ *                           2 (LSM_DISK_SMART_STATUS_GOOD):
+ *                              SMART Status good.
+ * @param[out] lsm_err
+ *                  Output pointer of lsm_error. Error message could be
+ *                  retrieved via lsm_error_message_get(). Memory should be
+ *                   by lsm_error_free().
+ * @return Error code as enumerated by \ref lsm_error_number.
+ * @retval LSM_ERR_OK               on success or not found.
```
Comment (tasleson) [inline c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h]: I'm trying to wrap my head around the `or not found`.  I see this is also present on some of the other calls, but at the moment it's not making sense to me?

File: c_binding/libsg.c [hunk 3]
```diff
@@ -940,3 +1035,385 @@ int _sg_host_no(char *err_msg, int fd, unsigned int *host_no)
...
+    *smart_status = _sg_smart_interpret_asc(asc);
+
+out:
+    return rc;
+}
+
+/*
+ * Query a SATA drive attached via SAS to get its SMART Status.
+ * 
+ * Input *smart_status should be a pointer to a one byte unsigned number.
```
Comment (tasleson) [inline c_binding/libsg.c]: Comment specifies `unsigned`, code is `signed`.

File: c_binding/libsg.c [hunk 4]
```diff
@@ -940,3 +1035,385 @@ int _sg_host_no(char *err_msg, int fd, unsigned int *host_no)
...
+ * The maximum *sd_name strlen is (_MAX_SD_NAME_STR_LEN - 1), assuming
+ * the caller did the check.
```
Comment (tasleson) [inline c_binding/libsg.c]: What is `sd_name` ?

Comment (tasleson) [inline c_binding/libsg.c]: Looks like documentation just needs to be updated :-)

File: c_binding/libsg.c [hunk 5]
```diff
@@ -940,3 +1035,385 @@ int _sg_host_no(char *err_msg, int fd, unsigned int *host_no)
...
+    assert(err_msg != NULL);
+    assert(fd >= 0);
+    assert(data != NULL);
+
+    memset(sense_err_msg, 0, _LSM_ERR_MSG_LEN);
+    memset(cdb, 0, _T10_SPC_LOG_SENSE_CMD_LEN);
+
+    cdb[0] = LOG_SENSE;
+    cdb[2] = (PAGE_CONTROL_CUMULATIVE_VALS << 6) | (page_code & 0x3f);
+    cdb[3] = sub_page_code & UINT8_MAX;
```
Comment (tasleson) [inline c_binding/libsg.c]: What's the purpose of `sub_page_code & UINT8_MAX` ?

File: c_binding/libsg.c [hunk 6]
```diff
@@ -940,3 +1035,385 @@ int _sg_host_no(char *err_msg, int fd, unsigned int *host_no)
...
+
+    assert(err_msg != NULL);
+    assert(fd >= 0);
+    assert(data != NULL);
+
+    memset(sense_err_msg, 0, _LSM_ERR_MSG_LEN);
+    memset(cdb, 0, _T10_SPC_LOG_SENSE_CMD_LEN);
+
+    cdb[0] = LOG_SENSE;
+    cdb[2] = (PAGE_CONTROL_CUMULATIVE_VALS << 6) | (page_code & 0x3f);
```
Comment (tasleson) [inline c_binding/libsg.c]: Should `PAGE_CONTROL_CUMULATIVE_VALS` value be (1 << 6) instead?

File: c_binding/libsg.h [hunk 7]
```diff
@@ -79,6 +79,12 @@
 #define _SG_T10_SPC_MODE_SENSE_MAX_LEN              0xffff
 /* ^ SPC-5 rev 12, 6.15 MODE SENSE(10) command introduction */
 
+#define _T10_SPC_LOG_SENSE_MAX_LEN                  0xffff
+/* ^ SPC-5 rev07 6.8 LOG SENSE command */
+
+#define _T10_SPC_REQUEST_SENSE_MAX_LEN              0xffff
+/* ^ SPC-5 rev07 6.35 LOG SENSE command */
```
Comment (tasleson) [inline c_binding/libsg.h]: `^ SPC-5 rev07 6.35 REQUEST SENSE command`

File: c_binding/libsg.h [hunk 8]
```diff
@@ -79,6 +79,12 @@
 #define _SG_T10_SPC_MODE_SENSE_MAX_LEN              0xffff
 /* ^ SPC-5 rev 12, 6.15 MODE SENSE(10) command introduction */
 
+#define _T10_SPC_LOG_SENSE_MAX_LEN                  0xffff
+/* ^ SPC-5 rev07 6.8 LOG SENSE command */
+
+#define _T10_SPC_REQUEST_SENSE_MAX_LEN              0xffff
```
Comment (tasleson) [inline c_binding/libsg.h]: I believe this value should be `0xFF` not `0xFFFF`

Comment (joehandzik) [inline c_binding/libsg.h]: @tasleson I don't disagree, but it seemed like precedent had been to set that command size for other commands that wouldn't theoretically require that much space either. I can change it and test to make sure it works.

File: c_binding/libsg.c (hunk 5 above)
Comment (joehandzik) [inline c_binding/libsg.c]: @tasleson This is another precedent thing. There are other cases where I saw this being used against single byte values (or so I thought). I'll scrub to see if I saw something wrong, and likely sanity test to make sure that it works without the masking.

File: c_binding/libsg.c (hunk 3 above)
Comment (joehandzik) [inline c_binding/libsg.c]: @tasleson Good catch...I thrashed a little bit on return values while I was working on this.

File: c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h (hunk 2 above)
Comment (joehandzik) [inline c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h]: @tasleson Stole this from surrounding function headers, I'll see if I can wordsmith better and if I cannot I'll remove it.

File: c_binding/libsg.c [hunk 9]
```diff
@@ -940,3 +1035,385 @@ int _sg_host_no(char *err_msg, int fd, unsigned int *host_no)
...
+ * 3. Return the asc/ascq values to the caller.
+ * 
+ * Input *asc should be a pointer to a two byte unsigned number.
+ * Input *ascq should be a pointer to a two bytes unsigned number.
+ * The maximum *sd_name strlen is (_MAX_SD_NAME_STR_LEN - 1), assuming
+ * the caller did the check.
+ * Return LSM_ERR_NO_MEMORY or LSM_ERR_NO_SUPPORT or LSM_ERR_LIB_BUG or
+ * LSM_ERR_NOT_FOUND_DISK.
+ */
+int _sg_sas_smart_status(char *err_msg, int fd, int32_t *smart_status)
```
Comment (tasleson) [inline c_binding/libsg.c]: The source documentation needs to be updated

**Comment by joehandzik** [comment]
@tasleson For the first round, fixed everything except for the UINT8_T masking and the REQUEST SENSE length. I'll update those after a quick test.

File: c_binding/libsg.c (hunk 5 above)
Comment (joehandzik) [inline c_binding/libsg.c]: @tasleson Example usage: https://github.com/libstorage/libstoragemgmt/blob/master/c_binding/libsg.c#L266

I'll leave this one until @cathay4t comments. From a code perspective, you're correct to say that it's not necessary. I'm indifferent.

File: c_binding/libsg.h (hunk 8 above)
Comment (joehandzik) [inline c_binding/libsg.h]: @tasleson I'll leave this one too until @cathay4t comments.

File: c_binding/libsg.c (hunk 5 above)
Comment (cathay4t) [inline c_binding/libsg.c]: Tony is right. There is no need to do `&UINT8_MAX` for a `uint8_t`.

File: c_binding/libsg.h (hunk 8 above)
Comment (cathay4t) [inline c_binding/libsg.h]: For `REQUEST SENSE`, the `ALLOCATION LENGTH` is 1 byte, which max length is 0xff.
I double checked all other XXX_MAX_LEN, they all fit into SCSI SPEC for 0xffff.

**Comment by cathay4t** [comment]
Please fix these `git am` warnings:

```
Applying: lsm_local_disk: Add SMART Status query method
.git/rebase-apply/patch:303: trailing whitespace.
    uint8_t count; 
.git/rebase-apply/patch:431: trailing whitespace.
        
.git/rebase-apply/patch:445: trailing whitespace.
        return -1;  
.git/rebase-apply/patch:481: space before tab in indent.
        if ((asc == 0) && (ascq == _T10_SPC_ASCQ_ATA_PASSTHROUGH)) {
.git/rebase-apply/patch:655: trailing whitespace.
    
warning: squelched 4 whitespace errors
warning: 9 lines add whitespace errors.
```

**Comment by cathay4t** [comment]
I got error on a hpsa raid disk:
```
>>> print lsm.LocalDisk.smart_status_get("/dev/sda")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/fge/vm/fge_Source/libstoragemgmt/python_binding/lsm/_local_disk.py", line 166, in smart_status_get
    return _use_c_lib_function(_local_disk_smart_status_get, disk_path)
  File "/home/fge/vm/fge_Source/libstoragemgmt/python_binding/lsm/_local_disk.py", line 33, in _use_c_lib_function
    raise LsmError(err_no, err_msg)
lsm._common.LsmError: LIB_BUG(1): Got error from SGIO LOG SENSE with error -1(Unknown error -1), Got SCSI sense data, key ILLEGAL REQUEST(0x05), ADDITIONAL SENSE CODE 0x20, ADDITIONAL SENSE CODE QUALIFIER 0x00, all sense data in hex: 700005000000000a00000000200000000000 
```

The output of smartctl:
```
[fge@megadeth ~]$ sudo smartctl -i /dev/sda -H
smartctl 6.5 2016-05-07 r4318 [x86_64-linux-4.8.15-200.fc24.x86_64] (local build)
Copyright (C) 2002-16, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Vendor:               HP
Product:              LOGICAL VOLUME
Revision:             2.14
User Capacity:        299,966,445,568 bytes [299 GB]
Logical block size:   512 bytes
Logical Unit id:      0x600508b1001c79ade5178f0626caaa9c
Serial number:        PDNNF0ARH6W07M
Device type:          disk
Local Time is:        Wed Apr  5 06:49:02 2017 EDT
SMART support is:     Available - device has SMART capability.
SMART support is:     Enabled
Temperature Warning:  Disabled or Not Supported

=== START OF READ SMART DATA SECTION ===
SMART Health Status: OK

```

**Comment by cathay4t** [comment]
Overall, I am quit impressed. The current code could actually identified a failed disk(I keep it just for this kind of testing).

**Comment by joehandzik** [comment]
@cathay4t Ah, the reason why the SMART status works in smartctl:

Years ago, a coworker of mine put code into smartmontools that actually leveraged the proprietary Smart Array APIs to pull out SMART data behind a RAID volume. The team responsible for ssacli was not and is still not happy that those APIs were used publicly. So, if we want SMART data behind a RAID volume, we'll want to put in a command for the plugins to extract that data from ssacli and storcli. So, I don't have any plans to make this command succeed against a RAID volume.

However, I'm wondering if we should try to return this as a STATUS_UNKNOWN, since this isn't really a bug but instead the device being targeted really can't support the functionality. Thoughts?

Otherwise, glad that it's working on your end too! This one took longer than expected, but now we have SCSI-to-ATA translation available in the code so future commands should take much less work.

**Comment by cathay4t** [comment]
@joehandzik I am OK to skip the proprietary API and  just return `LSM_ERR_NO_SUPPORT` and `LSM_DISK_SMART_STATUS_UNKNOWN` for hpsa RAID disks. We already have `lsm.disk.status` for disks behind HW RAID.

File: c_binding/libata.h [hunk 10]
```diff
@@ -22,7 +22,50 @@
...
+
+struct _ata_registers_output_28_bit {
+    uint8_t error;
+    uint8_t count;
+    uint8_t lba_low;
+    uint8_t lba_mid;
+    uint8_t lba_high;
+    uint8_t device;
+    uint8_t status;
+};
```
Comment (joehandzik) [inline c_binding/libata.h]: @cathay4t Are you sure? Everything I have seen (I thought) shows the fields being a single byte for ATA PASSTHROUGH (12): http://ftp.t10.org/ftp/t10/document.04/04-262r8.pdf

File: c_binding/libsg.c [hunk 11]
```diff
@@ -66,6 +70,26 @@
...
+#define _T10_SPC_INFO_EXCEPTION_PAGE_CODE                    0x2f
+
+/* SPC-5 rev07 Table 151 - Page control (PC) field */
+#define PAGE_CONTROL_CUMULATIVE_VALS                         0x01
+
+/* SPC-5 rev07 Table E.13 - Mode pages codes */
+#define INFO_EXCEPTION_CONTROL_PAGE                          0x1c
+
+/* Found documentation via Seagate SCSI Commands Reference Manual */
+#define MRIE_REPORT_INFO_EXCEPTION_ON_REQUEST                0x6
```
Comment (joehandzik) [inline c_binding/libsg.c]: @cathay4t Ok, I'll update that piece of documentation!

File: c_binding/libata.h (hunk 10 above)
Comment (joehandzik) [inline c_binding/libata.h]: @cathay4t I'll add the pack(push, 1) though, forgot to bring it over to libata.h (when I started this PR last year, libata.h didn't exist yet).

Comment (cathay4t) [inline c_binding/libata.h]: You are right. I was reading the ancient ACS-1(I always forgot to delete that PDF).
In both ACS-2 and ACS-3 "Normal and Error Output field descriptions", they are 1 byte.

**Comment by cathay4t** [comment]
@joehandzik Have you start any code work on NVMe SMART? I created issue #265 to trace this.

File: c_binding/libsg.h (hunk 8 above)
Comment (joehandzik) [inline c_binding/libsg.h]: So, just to be clear, I should actually change the REQUEST_SENSE_MAX_LEN value to 0xff?

**Comment by joehandzik** [comment]
@cathay4t Ok, I changed REQUEST_SENSE_MAX_LEN to 0xff from 0xffff preemptively. I also updated the code so ILLEGAL_REQUEST return values from LOG_SENSE and REQUEST_SENSE are interpreted as LSM_ERR_NO_SUPPORT rather than bugs. I believe that I have addressed your other PR comments and the whitespace issues you caught too. I chose to not change the naming of anything to the info_excep naming structure, mostly because I wasn't quite sure how far you wanted me to go with that. If you still want me to make that change, can you please comment on the specific methods and variables that need to be changed?

Other than that one change, we should be good to go. I figure any changes you want to make to support fibre channel can be made in a follow-on PR (even if they're trivial, since I'm not able to test FC at this time).

Thanks!

File: c_binding/libsg.c [hunk 12]
```diff
@@ -170,12 +218,52 @@ struct _sg_t10_sense_dp {
...
     uint8_t we_dont_care_0[4];
     uint16_t block_dp_header_len_be;
 };
 
+struct _sg_t10_log_para_hdr {
+    uint8_t we_dont_care_0[2];
+    uint16_t log_data_len_be;
+};
+
+struct _sg_t10_informational_exceptions_mode_page_0_hdr {
```
Comment (cathay4t) [inline c_binding/libsg.c]: If possible, please name this struct to `struct _sg_t10_info_excep_mode_page_0_hdr`.

Comment (joehandzik) [inline c_binding/libsg.c]: Updated, along with anything else that had 'informational_exceptions' or 'info_exceptions' in the name.

**Comment by cathay4t** [comment]
@joehandzik @tasleson Since SMART is a ATA term(SCSI use `informational exceptions`, NVMe use `SMART / Health Information`), should we use more generic name like `lsm_local_disk_health_status_get()`?

**Comment by tasleson** [comment]
@cathay4t @joehandzik I guess it would be good to keep the term hardware agnostic considering all the different storage types that are available out there.

**Comment by joehandzik** [comment]
@tasleson @cathay4t I will change accordingly sometime today, then. I'm fine with it, and it's more akin to the phrasing that Redfish has chosen for disk health.

**Comment by joehandzik** [comment]
@tasleson @cathay4t Updated the naming and stuck in a copyright update for the files I touched from the HPE side. Lemme know what you think, and if we're all good please feel free to merge! :)

**Comment by cathay4t** [comment]
@joehandzik Merged. I set the devel tree to 1.5.0 (46c0760aad11313e744494b04e11338fb1b64359) before merge this PR. Great hanks for this long wait function.

**Comment by joehandzik** [comment]
@cathay4t Awesome, thanks! Hopefully can iterate faster now that we have some ATA passthrough code to reuse!

**Pull request merged**